### DISCUSSION TITLE
XOR and Whitout<...> types in Prisma

### DISCUSSION THREAD

**Discussion by chrishoermann** [prisma/prisma#18283] [Q&A] [answered]
I've built a generator for prisma ([zod-prisma-types](https://github.com/chrishoermann/zod-prisma-types)) that throws a type error for every arg property that uses the `XOR` type. 

The problem I can't solve lies somwhere in the `Whitout<...>` type that I don't really understand. 

If I'm not mistaken the `Without<...>` type is used to extract properties that are not shared between types and are then merged back into the type where they don't exist. The XOR then creates a union of these types. Is this guess somwhere in the ballpark? 

The question then would be if this information can be reflected in more detail in the `dmmf` so generators can access this information to generate e.g. the right zod types. Because now there is no way that I know of that would let me generate the right type with zod.

To try out what I mean I've put together a test repo: https://github.com/chrishoermann/zod-prisma-types-bug

**Comment by chrishoermann** [comment, marked as answer]
Actually this seems to be a `zod` issue but nontheless I'm interested in how the Without type works exactly

**Reply by nurul3101** [reply]
Hello @chrishoermann 👋 

From the generated prisma client, I can see this `Without` type generated. I assume you want to understand the functioning of this type?

<img width="629" alt="Screenshot 2023-03-21 at 3 39 13 PM" src="https://user-images.githubusercontent.com/13049676/226575152-57953e0b-9d3f-4749-b91c-0174e59dee66.png">

This `Without` type is a utility type in TypeScript that creates a new type by setting specific properties from an existing type (T) to never. The properties to be set to never are determined by excluding the keys from another type (U)

**Reply by nurul3101** [reply]
@chrishoermann Do let us know in case you have any further questions.
